kuma/api/tasks.py

```diff
@@ -114,3 +143,50 @@ def publish(doc_pks, log=None, completion_message=None):
 
     if completion_message:
         log.info(completion_message)
+
+    cdn_cache_invalidate.delay(doc_locale_slug_pairs)
+
+
+@task
+def cdn_cache_invalidate(doc_locale_slug_pairs, log=None):
+    """
+    Independent of a document being updated, added, or removed, trigger
+    an attempt to purge it from the CDN.
+    """
+    if not log:
+        log = cdn_cache_invalidate.get_logger()
+
+    client = get_cloudfront_client()
+    for label, conf in settings.MDN_CLOUDFRONT_DISTRIBUTIONS.items():
+        if not conf['id']:
+            log.info('No Distribution ID available for CloudFront {!r}'.format(
+                label
+            ))
+            continue
+        paths = [
+            conf['transform'](locale, slug)
+            for locale, slug in doc_locale_slug_pairs
+        ]
+        # In case the transform function decided to "opt-out" on a particular
+        # (locale, slug) it might return a falsy value.
+        paths = [x for x in paths if x]
+
+        if paths:
+            invalidation = client.create_invalidation(
```

Comment:
is there any way to throttle invalidations? some of our API endpoints get hit with an unreasonable amount of requests

Reply:
Uh? What endpoints? 

Throttling all these cache invalidation requests might not make sense. If you edit an individual wiki document, and this doesn't happen very often, it will trigger one little request like this. 

If you run the `./manage.py publish --all` command it will send 1,000 of these per request but the limit is 3,000. At least that's how I read it. 3,000 per request is the max. Not 3,000 per hour or anything like that.

Reply:
> is there any way to throttle invalidations?

never mind, the worf autoblocker will automatically throttle any endpoints. 

Couldn't the `./manage.py publish --all` command invalidate `/foo/*`?

Reply:
From my reading, it's not 3000 per request, but 3000 in-play at any given time (see https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Invalidation.html#InvalidationLimits):

> If you're invalidating files individually, you can have invalidation requests for up to 3,000 files per distribution in progress at one time.

Which we would quickly exceed when doing mass re-renders of documents (e.g., when I re-render all of the documents that use the `compat` macro, which I do every few weeks). So in my view, we have to somehow distinguish between the two modes, occasional edit/move/delete/re-render events, and mass events, and handle each one differently. For mass events, we'll have to use the wildcard invalidation approach instead since they have no file limit (we're allowed up to 15 wildcard invalidations in-play at any time).